include/envoy/network/io_handle.h

```diff
@@ -179,18 +179,33 @@ class IoHandle {
   virtual Api::SysCallIntResult getOption(int level, int optname, void* optval,
                                           socklen_t* optlen) PURE;
 
-  /**
-   * Get local address to which handle is bound (see man 2 getsockname)
-   */
-  virtual Api::SysCallIntResult getLocalAddress(sockaddr* address, socklen_t* addrlen) PURE;
-
   /**
    * Toggle blocking behavior
    * @param blocking flag to set/unset blocking state
    * @return a Api::SysCallIntResult with rc_ = 0 for success and rc_ = -1 for failure. If the call
    * is successful, errno_ shouldn't be used.
    */
   virtual Api::SysCallIntResult setBlocking(bool blocking) PURE;
+
+  /**
+   * Get domain used by underlying socket (see man 2 socket)
+   * @param domain updated to the underlying socket's domain if call is successful
+   * @return a Api::SysCallIntResult with rc_ = 0 for success and rc_ = -1 for failure. If the call
+   * is successful, errno_ shouldn't be used.
+   */
+  virtual Api::SysCallIntResult domain(int& domain) PURE;
```

Comment:
Given the failure modes of `getsockname` (only fails on bad fd, or kernel resources exhausted) and how the error is handled, I think the interface to this would be nice to just return `absl::optional<int>` and take no arguments.

Reply:
You're right, looks better! Done!